### ISSUE TITLE
Accept pytorch Dataset as input on pipeline training method

### ISSUE THREAD

**Issue by frascuchon** [recognai/biome-text#294]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**

Accepting standard pytorch Dataset as input in pipeline training method could be more flexible integration with another data preparation tools or already created datasets, like [huggingface nlp datasets](https://github.com/huggingface/nlp)

The Pipeline could adds a new method for dataset creation from pipeline configuration, vocabulary and the selected datasources: `Pipeline.create_dataset(ds)`. And could include actions of training-validation-test split or some metrics results.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by frascuchon** [comment]
# Pytorch dataset integration

## Introduction

Using torch dataset as input in `Pipeline.train` method, we try to find library to standard or common components. But there are some key points to address before all those method changes that we will discusse in following steps.

### torch.Dataset is just an api definition

The base `torch.utils.data.dataset.Dataset` class is an abstract class specification and no exists default implementations. On the one hand, this could be an advantage for better input training dataset generalization, but the truth is that allennlp training proccess is hard coupled to dataset internal structure: input data must be transformed to `allennlp.data.Instance` object, then those instances are collected as torch tensors when dataset is chunked in batches.

### Allennlp doesn't clear separate features generation pipeline

Pipeline model forward method accepts a tensor dictionary representing token indexes at feature level (word or char), but input data is representing using `allennlp.data.Instance`, keeping this tensor transformation as an internal and convoluted process involving instance, batch and vocabulary components.

### Allennlp datasets are used for both, pipeline training and vocabulary creation

Before the train process, every pipeline must be configured using a `allennlp.data.Vocabulary` that normally is built using training data instances. So, multiples iterations are applied over instance data generation (at least, one for vocab creation and one for training).

## Solution Approach

Allennlp wraps data instances in a torch DataSet implementation (`AllennlpDataset` and `AllennlpLazyDataset`). We can exploid this implementation, and use it as input data point for out `Pipeline.train` method. 

We can bring methods for easily build a datasource/dataframe or whaterver data collection into instance Datasets. These instance datasets will be used for vocabulary creation and pipeline training, without "re-generate" instances in training phase.

Also, depending on source data size, we could decide generate instances dataset lazily or not. This will impact into torch `Dataloader` features (`IterableDataset`s are not "bucketizables", for example).

A code snippet describing the flow could be:

```python
pipeline = Pipeline.from_config(....)
# Instancd torch Dataset creation
train_dataset = pipeline.create_dataset(training_data,....) # could have extra arguments
validaiton_dataset = pipeline.create_dataset(validation_data,...)
# Fit the pipeline vocabulary using the datasets
pipeline.create_vocabulary(VocabularyConfiguration(sources=[traind_dataset, validation_dataset])
# Then, train pipeline with same instance datasets
pipeline.train(trainer_cfg, training=train_dataset, validation=validation_datataset)
```
